File: LeetCode/Python/solutions/p2439.py
import math
from typing import List


class Solution:
    def minimizeArrayValue(self, nums: List[int]) -> int:
        #
        # Solution v1: MovingAverage
        #
        # This equal to find MovingAverage in order.
        # In other word, smaller element can not affect the average of previous sub
        # Thus, this question is equivalent to finding the maximum substring average
        #
        # Runtime: 840 ms @ (beats) 54.71%
        # Memory Usage: 26 MB @ (beats) 99.55%
        #
        ans = nums[0]
        sub_total = nums[0]

        for i in range(1, len(nums)):
            sub_total += nums[i]
            ans = max(ans, math.ceil(sub_total / (i + 1)))

        return ans


if __name__ == "__main__":
    solution = Solution()

    nums = [3, 7, 1, 6]
    print(f"return: {solution.minimizeArrayValue(nums)}")  # 5

    nums = [10, 1]
    print(f"return: {solution.minimizeArrayValue(nums)}")  # 10

    nums = [6, 9, 3, 8, 14]
    print(f"return: {solution.minimizeArrayValue(nums)}")  # 8

File: LeetCode/Java/src/solutions/p2439_MinimizeMaximumofArray/P2439.java
package solutions.p2439_MinimizeMaximumofArray;

public class P2439 {
    public int minimizeArrayValue(int[] nums) {
        /*
         * Solution v1: Mving Average
         *
         * Runtime: 7 ms @ (beats) 97.13%
         * Memory Usage: 62 MB @ (beats) 46.24%
         */
        double ans = nums[0];
        double sub_total = nums[0];

        for (int i = 1; i < nums.length; i++) {
            sub_total += nums[i];
            double avg = Math.ceil(sub_total / (i + 1));
            if (avg > ans) {
                ans = avg;
            }
        }
        return (int) ans;
    }
}
